### DISCUSSION TITLE
Possibili miglioramenti

### DISCUSSION THREAD

**Discussion by VegethB** [MainKronos/Sonarr-AnimeDownloader#15] [Ideas]
Come suggerito da @MainKronos, ecco una lista di altre features:
(**NON SONO IN ORDINE DI PRIORITA'!** _Anzi, non ci sono proprio priorità, è solo una lista di features_)

## Main App:
- [ ] Possibilità di scegliere   una data/un range di tempo   specifico per le scans (un jobs scheduler). _Molte serie sono di VVVVID/Crunchyroll o comunque vengono rilasciate massimo con 1 giorno di ritardo (speedSubs), trovo che sia poco utile far partire la scan ogni X minuti/ore 24/7 quando basterebbe pianificare la scan per quella specifica serie ad un intervallo di date._ Questo porta a molti vantaggi:
 1. Si riduce il carico di lavoro sulla CPU... Ora, io uso un Dual CPU Xeon quindi sai che differenza mi fa... discorso diverso per chi usa HW come Arduino / NAS del paleolitico ancora venduti come super macchine del futuro _(una mia conoscenza usa un NAS del 2012 e fa fatica a tenere in piedi pure l'interfaccia Web QNAP, nonostante questo, può usare docker proprio al minimo ma anche così esplode tutto 😂)_;
2) Minor numero di richieste al server AnimeWorld = Minor carico di rete _(nel 2021 in Italia ancora si usa il "doppino a 54kb" se non sei in città 🤢) e magari si evita anche la paranoia del ban ip dal server di animeworld (🤷‍♂️)_;
3) Log decisamente più pulito e leggibile _(Docker Desktop, poi non so su linux ecc., mostra il log fino ad un tot di righe... quelle più vecchie vengono cancellate)._ Se si riduce il numero di volte che parte una scan destinata a fallire è meglio;
- [ ] Poter usare altre source per il download dei vari episodi (VVVVID per esempio). Questo perché è già stato appurato che AnimeWorld e molti altri siti _Warez_ Ri-Encodano la fonte video originale (come per i video vvvvid) e perché non è detto che AnimeWorld abbia certe serie. Esempio: Nanatsu no Taizai S04 episodio 6 Fonte originale: 596.1MB VS 325MB di AnimeWorld. 
Se fosse troppo complicato si può integrare [AnimeGG](https://gitlab.com/animegg/animegg) o [Youtube-DL](https://github.com/ytdl-org/youtube-dl/) (quest'ultimo usato da [AnimeGG](https://gitlab.com/animegg/animegg)) per scaricare le serie da VVVVID;
- [ ] Beh...! Se abbiamo il supporto ad AnimeWorld perché non aggiungerlo anche a HentaiWorld? 😂😁
_Effettivamente è inutile dato che Sonarr usa (per ora) esclusivamente TheTVDB il quale VIETA serie +18. Però non sarebbe male pensare al futuro, quando si decideranno a droppare TheTVDB e supportare AniDB/AniList/MAL ecc.;
Magari si potrebbe ri-creare un mini Sonarr (qui) che tramite API di AniList/MAL/AniDB ecc. (se necessario anche usando un account proprio, come per telegram) si possono aggiungere le serie +18. Le quali verranno salvate in una tabella delle serie "monitorate" dal programma (non da sonarr). A quel punto, nella tabella JSON si aggiungerà il nome della serie monitorata e il solito link di HentaiWorld (che è identico in apparenza a quelli usati da AnimeWorld). Per le serie gestite dal programma verrà quindi richiesto di Segnalare la posizione della cartella che conterrà tutte le serie di HentaiWorld. Il nome della serie sarà lo stesso di quello monitorato (includendo il fix di conversione delle lettere invalide come `:` in `_`);_

## Sezione logs:
- [ ] Scrivere la versione attuale del programma nel Log di output;
- [ ] Riferito a 
> - [ ] Possibilità di scegliere   una data/un range di tempo   specifico per le scans (un jobs scheduler).

Aggiungere l'output: Nome Jobs - tipo Jobs - avvio/risultato inserendo all'interno i risultati di output del tipo di jobs (esempio:
```
Job1 - scan pianificata - avvio...
soliti output dello scanner per le serie interessate
Job1 - scan pianificata - fine, nessun errore
```
(errore relativo allo schedulatore);
- [ ] Far si che si possa usare un Log di output in modalità di debug nel quale **NON** vengano scritte in modo chiaro informazioni sensibili (esempio le api key). Questo per far si che si possa usare/copiare il log con "CTRL + C" "CTRL + V" su Github, permettendo di condividere il log stesso per gli errori (senza dover cancellare o nascondere in qualche modo le informazioni sensibili). _Non ricordo se è questa docker o un'altra, aggiungere il log da poter condividere alle mount o nella cartella dei DATI_;
- [x] ~~Aggiungere nel log di output il tempo impostato per il parametro "Intervallo Scan"~~;
- [ ] Mostrare il nome originale del file quando viene scritto che è stato rinominato (può aiutare in casi di problemi dovuti al nome originale... tipo nome troppo lungo, caratteri speciali o altro;
- [ ] Oltre a mostrare l'errore (con debug attivo), aggiungere un link alla issue su github se è un errore conosciuto ma non ancora risolto (_inoltre aprire una issue con l'errore conosciuto è una cosa che vedo molto spesso fare dagli autori_);

## Pagina Web (WebUI - JSON):
- [ ] Scrivere la versione attuale del programma nella pagina Web (in settings o altrove o in una pagina dedicata insieme ad altre informazioni utili);
- [ ] In "intervallo scan" nelle impostazioni, mostrare il tempo in un riquadro invece che su popup a comparsa spostando lo slide con il mouse;
- [ ] Mettere un menu dove poter cambiare pagina tramite click e quindi poter passare dalla home alle settings e viceversa. In questo momento, per tornare indietro va usato il tasto back del proprio browser web... in certi casi è confusionale perché non sai se salva o meno i cambiamenti apportati;
- [ ] Possibilità di modificare le serie già presenti. Se decidessi di monitorare 4 stagioni di una serie e sbaglio a mettere il link della 4 stagione, devo rifare tutto da 0 (su interfaccia web)... non è carino 🙃;
- [ ] Come detto nella discussione sui pareri dopo 1 settimana (#11 ), Aggiornare la pagina con un layout (non forzato, quindi a scelta del utente finale) di visualizzazione a Card, le quali mostreranno la locandina della serie (presa dal link di AnimeWorld) e le stagioni monitorate (più una icona o qualcosa che indichi se ci sono stati errori per quella specifica serie, tipo ultimo episodio trovato ma non scaricato per qualche motivo);
- [ ] Possibilità di rimandarti a Sonarr (per quella specifica serie) tramite iconcina (quando espandi il menu);
- [ ] Pagina dedicata alla visualizzazione del LOG con possibilità di scaricarlo (al momento è possibile vedere il log SOLO dalla Docker... Scomodissimo)

## **EXTRA** Sistemi:
Questa lista è solo a scopo informativo. Se non è possibile convertire il progetto per questi sistemi e architetture, si tolgono dalla lista (questo SOLO se è proprio impossibile, tipo libreria X solo a 64bit solo per OS Linux).
_È vero, docker tramite macchina virtuale va a """eliminare""" questo problema (creandone altri 100 ma dettagli...) tuttavia è frustante per molti dover usare docker (e non tutti i NAS ancora in giro supportano la virtualizzazione o sono abbastanza potenti da usare docker in contemporanea) quando sia Sonarr che Plex/Emby sono eseguibili come software diretto per quei specifici OS._

- [ ] Supporto a Windows tramite EXE/MSI, CPU con segmento desktop (visto che ora esiste Windows per ARM, specifichiamo);
- [ ] Supporto diretto al sistema linux (niente docker in poche parole) CPU con segmento desktop;
- [ ] Supporto per CPU con segmento Mobile (ARM ecc.)

### Esempio pratico dei danni causati da Docker per un utente Windows:
1. Dal 10% al 30% di risorse CPU + RAM perse per far funzionare Hyper-V in background (e nel mio caso ho un dual CPU Xeon ma si fa sentire comunque 🤬); 
2. Siccome alla Microsoft WSL faceva cagare, hanno fatto WSL2 che fa ancora più schifo ma almeno risolve un grande problema (SOLO SE USI LE BUILD INSIDER... quindi solo se sei un folle pazzo scatenato ubriaco), quello di poter usare WSL2 al posto di Hyper-V. Perché tutto questo? Perché Hyper-V è una 💩💩💩 e tutti usano VMware ma VMware non funziona se hai il servizio Hyper-V attivo (a meno che non disabiliti certe sicurezze in Windows... cosa SCONSIGLIATTISSIMA);
3. Devi imparare ad usare non solo Docker e i suoi comandi (perché la GUI è una super 💩💩💩💩💩💩... puoi mettere le mount ma non le variabili ENV... 🙃🤷‍♂️🤬) ma anche Linux! Fidati che le bestemmie sono dietro al primo angolo;
4. Io non capisco dove sia la semplicità su Docker... Se qualcosa non va e non sai nemmeno cos'è linux, sei totalmente fregato. Essendo tu utente Windows per quale motivo dovresti conoscere i comandi linux o sapere che cosa è la shell/CLI?
Se volessi cambiare la config del software dove è finito il mio .ini / .xml? Sono tutte cose che già su Ubuntu fai fatica ma grazie a una interfaccia grafica puoi girare come su windows e far qualcosa (ma soprattutto **puoi accedere/copiare/spostare** i file come faresti su windows... cosa praticamente impossibile visto che Dokcer è più blindato di un carrarmato). Per farti capire, il comando che TUTTI mettete dando per scontato che siamo tutti studiati di Docker (OVUNQUE STA COSA!)
![image](https://user-images.githubusercontent.com/17975351/109025068-af30b200-76be-11eb-9c78-72ff1ab71938.png)
NON FUNZIONA SU WINDOWS!!!
Perché va scritto così:
`docker run -d --name=NomeCheVuoiTU -v B:\SERVER\DOCKER\AnimeDownloaderSR\DATA:/script/json/ -v LetteraUnità\path\to\folder1\:/Anime -v LetteraUnità\path\to\folder2\:/Anime2 -p 5000:5000 --env SONARR_URL='IP SONARR' --env API_KEY='TI PIACEREBBE 😃' --env TZ=Europe/Rome mainkronos/anime_downloader`
Questo già fa capire perché ODIO docker.
NON È FINITA! Docker dovrebbe essere Easy perché per aggiornare la dock con il nuovo update basta fare `pull`... Un corno! Devi fermare la dock, cancellarla, allora fai il pull e poi su powershell inserisci il tuo comando (quello sopra) quando su windows fa tutto in automatico il programma o se usi python ti basta scaricare il master e sovrascrivere i file (se non ci sono file di config premade), cosa che molti scripts come SickChill fanno già in automatico 🤷‍♂️.

**Comment by MainKronos** [comment]
Dovrei essere riuscito a fare un immagine multi piattaforma compatibile con arm64, amd64, arm
Non sò se intendevi questo nella sezione EXTRA Sistemi😅

**Reply by VegethB** [reply]
> Non sò se intendevi questo nella sezione EXTRA Sistemi😅

Intendevo dire questo:
Utente Windows, apro github (questo), nei download (release) troverò un .exe,
aprendo questo .exe verrà installato il programma e creata una icona sul desktop che aprirà la pagina web per "gestire" il backend (quella amministrativa).

È un esempio (usando win come esempio dovrebbe essere più facile capire), potrà essere anche un .msi o altro.
Il succo è che deve essere un formato eseguibile per quel SO (windows) e non una Docker che mi emuli linux.
Inoltre questo punto è un extra, una feature da fare dopo tutto il resto 👍.


> Dovrei essere riuscito a fare un immagine multi piattaforma compatibile con arm64, amd64, arm

Qui siamo fermi alla 1.3.2 😂

**Comment by berta24** [comment]
Un informazione, io utilizzo Unraid come OS per gestire i docker, ed esso ha una funzione di update dei vari container
![unraid](https://user-images.githubusercontent.com/25983005/118820726-d3afbb00-b8b6-11eb-901d-21aa123fd983.png)
é possibile implementare questa cosa? oppure è troppo di nicchia e non ne vale la pena?

**Reply by MainKronos** [reply]
Non ho ben capito che cosa vorresti che ci fosse... 😅
Un sistema di aggiornamento automatico?

**Reply by berta24** [reply]
Semplicemente un modo di vedere ad occhio sulla pagina dei docker la versione, se è disponibile un aggiornamento oppure se è up to date, 
![Immagine 2021-05-19 153545](https://user-images.githubusercontent.com/25983005/118821850-eaa2dd00-b8b7-11eb-85cb-02115968734d.png)
Tipo così

**Reply by MainKronos** [reply]
ah! interessante....
vedo se poi riesco a farlo 👍
